### DISCUSSION TITLE
Nernst-Planck & charge conservation

### DISCUSSION THREAD

**Discussion by pbriml** [PMEAL/OpenPNM#2647] [Q&A]
Hello, I am working to develop a pore network model that uses the discretized Nernst-Planck and charge conservation equations (still deciding between Poisson's or electroneutrality) as described in this work: 

> Mehrez Agnaou, Mohammad Amin Sadeghi, Thomas G. Tranter, Jeff T. Gostick,
> Modeling transport of charged species in pore networks: Solution of the Nernst–Planck equations coupled with fluid flow and charge conservation equations, Computers & Geosciences, Volume 140, 2020, 104505, ISSN 0098-3004, https://doi.org/10.1016/j.cageo.2020.104505.
> 

In v3 of OpenPNM, it does not appear that there is a pre-built algorithm for Nernst-Planck charge transport or charge conservation. 

**Do the developers have plans to add this functionality anytime soon, or do they have recommendations for ways to do this manually? Any tips/tricks are appreciated!** 

I was thinking about perhaps doing a manual definition of conductance as is done similarly [in this example](https://openpnm.org/examples/tutorials/08_simulating_transport.html): 

```
R = pn['throat.diameter']/2
L = pn['throat.length']
mu = water['throat.viscosity']  # See ProTip below
water['throat.hydraulic_conductance'] = np.pi*R**4/(8*mu*L)
print(water['throat.hydraulic_conductance'])
```

And then coupling with one of the transport algorithms? Would this idea also be suitable for charge conservation? 

Thanks!

**Comment by ma-sadeghi** [comment]
Hi @pbriml. We've decided to keep the built-in algorithms at a minimum, and instead provide more "example notebooks". The main reason is that these algorithms are often times very complex, which makes code maintenance very difficult. If you're interested, we'd more than welcome a pull request on creating an example notebook for a simple Nernst Planck implementation. The functionality is there, we just removed the pore-scale models specific to this particular algorith, which you can easily pull from an earlier version/tag.

**Reply by pbriml** [reply]
Thank you for the prompt response, I have created a new pull request. Your reasoning about the algorithms makes quite a lot of sense, thanks in advance for your help.

**Comment by pbriml** [comment]
Hello, I have a few questions that I am hoping one of the devs may be able to answer. I have been working to adapt [this script](https://github.com/PMEAL/OpenPNM/blob/dev/scripts/example_nernst_planck2d.py) (example_nernstplanck_2d.py) to be compatible with OpenPNM v3. So far, I have been able to make decent progress and have adapted the following files: 

- op.models.physics.ionic_conductance
- op.models.physics.ad_dif_mig_conductance
- op.algorithms.IonicConduction
- op.algorithms.NernstPlanck
- op.algorithms.NernstPlanckMultiphysicsSolver
- op.models.physics.source_terms 

I am, admittedly, not much of a developer and all of my research thus far has been focused on scientific computing rather than development (i.e. workflows, multiphysics simulations in Comsol & python, DFT calculations, post-processing, etc.). The way I've been calling all these edited files is by importing them into the simulation rather than the built-in OpenPNM packages. 

Currently, I have the following issue occur when I try to run the main file (2D_Nernst_Planck_recreation.py): 
```
`Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_bundle/pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Nernst-Planck/2D_Nernst_Planck_recreation.py", line 177, in <module>
    pnp.run()
  File "/Nernst-Planck/_NernstPlanckMultiphysicsSolver_pb.py", line 113, in run
    p_alg._charge_conservation_eq_source_term(e_alg=e_alg)
  File "/Nernst-Planck/_ionic_conduction.py", line 56, in _charge_conservation_eq_source_term
    phys[0].add_model(propname='pore.charge_conservation', model=mod,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/openpnm/core/_models.py", line 404, in add_model
    self.run_model(propname+'@'+domain)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/openpnm/core/_models.py", line 502, in run_model
    self.run_model(propname=propname, domain=domain)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/openpnm/core/_models.py", line 521, in run_model
    vals = mod_dict['model'](self, **kwargs)
TypeError: charge_conservation() got multiple values for argument 'phase'
```

When debugging, I have found that the issue occurs in the IonicConduction algorithm when trying to add the pore scale charge conservation model to the simulation: 

```
`class IonicConduction(ReactiveTransport):
    r"""
    A class to enforce charge conservation in ionic transport simulations.

    Parameters
    ----------
    network : GenericNetwork
        The network on which this algorithm operates
    name : str, optional
        A unique name to give the object for easier identification.  If not
        given, one is generated.
    """

    def __init__(self, name='_ionic_?', **kwargs):
        super().__init__(name=name, **kwargs)
        # self.settings = SettingsAttr(IonicConductionSettings, settings)
        self.settings._update(IonicConductionSettings())

    def _charge_conservation_eq_source_term(self, e_alg):
        # Source term for Poisson or charge conservation (electroneutrality) eq
        phase = self.project[self.settings.phase]
        Ps = (self['pore.all'] * np.isnan(self['pore.bc.value'])
              * np.isnan(self['pore.bc.rate']))
        mod = st.charge_conservation
        # phys = self.project.find_physics(phase=phase)
        phys = self.project.phases
        phys[0].add_model(propname='pore.charge_conservation', model=mod,
                           phase=phase, p_alg=self, e_alg=e_alg,
                           assumption=self.settings['charge_conservation'])
        self.set_source(propname='pore.charge_conservation', pores=Ps)
```

Specifically, in these lines: 

```
        phys[0].add_model(propname='pore.charge_conservation', model=mod,
                           phase=phase, p_alg=self, e_alg=e_alg,
                           assumption=self.settings['charge_conservation'])
```

I have done my best to adapt the attached files to OpenPNM v3, but I am really stuck on this error and am hoping for some guidance. I am also happy to send a more detailed description of the problem &/or my set-up in an email if that's more convenient. 

Thanks! 
[Nernst-Planck.zip](https://github.com/PMEAL/OpenPNM/files/10541082/Nernst-Planck.zip)

**Reply by jgostick** [reply]
I'm going to basically say that OpenPNM 3.0 does not support Nernst-Plank, and it could be a while before it does. If you really need this, then you'll need to stick with 2.X.  Sorry.  We are rethinking/redesigning how multiphysics algorithms are handled and we want to get it right, so it's slow.
